Feature/track password changes

miscampbell opened the pull request:
Please see asana card below:

https://www.pivotaltracker.com/story/show/152469384

Commit:
Adding logic to ensure the LastPasswordChangeDate is always updated.

Commit:
Adding password expiration interval setting

Commit:
Adding logic to detect when a users login has expired and throw an exception

Commit:
Adding passwordExpirationDaysInterval

Commit:
Adding logic to detect when a Password has now expired

Commit:
Adding validation check that the LastPasswordChangedDate is a valid date time object

Commit:
Adding logic to store a users previous passwords.
A setting has been added to the constructor to control this logic

Commit:
Adding logic to enable logging login attempts

Commit:
Adding unit tests for LoggedLoginProvider

Commit:
Adding logic to validate if a user has reused a previous password when it is changed.

Commit:
Adding logic to disable a users account for a configurable period of time.

https://www.pivotaltracker.com/story/show/152469461

acuthbert commented on src/User.php:
```diff
@@ -213,6 +228,27 @@ protected function getConsistencyValidationErrors()
+        //  Validate new password has not been previously used
+        if ($this->passwordChanged && AuthenticationSettings::singleton()->compareNewUserPasswordWithPreviousEntries) {
+
+            $numberOfPastPasswordsToCompareTo = AuthenticationSettings::singleton()->numberOfPastPasswordsToCompareTo;
```
I think we could maybe combine the flag setting and numerical setting into one - with zero being disabled. I think it might confused people who just discover the numerical setting, set it but don't see the feature work.

miscampbell replied:
This setting has now been removed and replaced with the property numberOfPastPasswordToCompareTo to control both scenarios

acuthbert commented on src/Settings/AuthenticationSettings.php:
```diff
@@ -27,4 +27,64 @@ class AuthenticationSettings extends Settings
+     *
+     * @var bool
+     */
+    public $disableAccountAfterFailedLoginAttempts = false;
```
I think these bools could be combined with their numerical setting - although I'd appreciate your thoughts on whether that is clean or not. With disabling I could see this as a 'tri' state setting, off, 'temporary' or 'permanent' where permanent requires a password reset to get going again. So in that case we would need the two types anyway.

acuthbert commented on src/Settings/AuthenticationSettings.php:
```diff
@@ -27,4 +27,64 @@ class AuthenticationSettings extends Settings
+     * The following is used to control whether a users previous password should be retained
+     * @var bool
+     */
+    public $storeUserPasswordChanges = false;
```
I'm not sure we ever want to store passwords unless we're using them for excluding previous passwords. Perhaps then we could scrap this and only use the $compareNewUserPasswordWithPreviousEntries setting.

miscampbell replied:
Hi Bert 

I do agree with you regarding the numerous boolean settings for controlling each set of behaviour. I am happy to change this and simply use the numeric value and when it is set to zero the feature will not run.

I will just have very notes on the property inside the AuthenticationSettings class

miscampbell replied:
I am happy to remove all the booleans for Password logging and simply rely on both settings and if they are greater than zero then the code behind them is executed. The only headache with this is that the user will need to ensure both settings are set to greater than zero.

* $totalPreviousPasswordsToStore
* $numberOfPastPasswordsToCompareTo

miscampbell replied:
I have updated the code to use the property compareNewUserPasswordWithPreviousEntries. This property is also set via the constructor for the Scaffold Authentication module.

acuthbert commented on src/LoginProviders/LogLoginAttemptsTrait.php:
```diff
@@ -0,0 +1,54 @@
+        } catch (LoginFailedException $loginFailedException) {
+            $this->createFailedUserLoginAttempt($username, (string) $loginFailedException);
+            throw $loginFailedException;
+        } catch (LoginExpiredException $loginExpiredException) {
```
Hi Michael, I'm missing something - these exceptions don't exist on this branch and I'd expect to see something in ModelLoginProvider/LoginProvider that throws them too... missing commits?

miscampbell replied:
Hi Bert 
These exceptions are defined within the main Rhubarb module I just followed the same pattern and have a separate PR in for the Rhubarb module that includes these new Exception. 

Please see link below:
https://github.com/RhubarbPHP/Rhubarb/pull/55

acuthbert commented on src/DatabaseSchema.php:
```diff
@@ -27,5 +27,7 @@ public function __construct($version = 0.1)
         parent::__construct($version);
 
         $this->addModel('User', User::class);
+        $this->addModel('UserPastPassword', UserPastPassword::class);
```
I think you use "Previous" a lot in other places as opposed to "Past". I prefer previous as it's clearer - could you refactor this easily?

miscampbell replied:
Hi Bert that is no problem I will refactor this.

miscampbell replied:
This model have now been changed with the name User Previous Password as I agree the terminology fits better with other descriptions that I have used.

acuthbert commented on src/DatabaseSchema.php:
```diff
@@ -27,5 +27,7 @@ public function __construct($version = 0.1)
 
         $this->addModel('User', User::class);
+        $this->addModel('UserPastPassword', UserPastPassword::class);
+        $this->addModel('UserLoginAttempt', UserLoginAttempt::class);
```
I will chat to you about this one - there is a need to get wider user auditing in place for ProjectBlue - however I'm not sure how something 'generic' is plugged into a specific module like this one. Perhaps it will be a more general name of tblAuthenticationUserAudit, a string 'type', a 'data' field and then we can store password changes, log in and log outs, and other actions like 'user changed this' kind of stuff.

Commit:
Removing compareNewUserPasswordWithPreviousEntries boolean property and instead am using the integer property $numberOfPastPasswordToCompareTo.

This flag is now responsible for turning the feature on and controlling how many previous password are checked before continuing

Commit:
Removing property storeUserPasswordChanges and have replaced with compareNewUserPasswordWithPreviousEntries as there should never be a scenario when the user should store passwords that they have no intention of checking against.

Commit:
Changing the name of the model UserPastPassword to UserPreviousPassword
Updated the property $numberOfPastPasswordsToCompareTo to $numberOfPreviousPasswordsToCompareTo to tie in with the terminology of previous.

Commit:
Changing the name of the model UserPastPassword to UserPreviousPassword
Updated the property $numberOfPastPasswordsToCompareTo to $numberOfPreviousPasswordsToCompareTo to tie in with the terminology of previous.

acuthbert committed:
WIP

acuthbert committed:
Refactord password change to work inside the provider

acuthbert committed:
Real world testing

acuthbert committed:
Unit tests passing.

acuthbert committed:
Real world testing

acuthbert committed:
Updated change log

acuthbert committed:
Supported changing identity column name within the model itself

acuthbert committed:
Supported changing identity column name within the model itself

Commit:
Fix for logic that validates the previously used password to validate against the password that was entered by the user rather than the one that is on the Users Model

acuthbert commented:
Hey Michael - this PR was originally for Somervilles - is it using your branch directly currently or is it back on normal tags - in other words - does this still need merged?

miscampbell commented:
Hi @acuthbert 

No I dont believe this needs merged anymore. Somervilles uses only tags and does not reference any specific commits or branches.
